### PR TITLE
Dark mode support

### PR DESCRIPTION
**Pull request by zwaldowski**
Completes #463.

### TODO

- [x] Investigate `controlBackgroundColor`.

---

This PR brings the dark mode implementation from @saagarjha and @douglashill into upstream.

In addition to their (awesome) work, this PR:

- Better supports the color scheme changing at runtime
- Makes some icon buttons light up instead of darkening when pressed in Dark Mode
- Makes rows in the configuration editor are readable in Dark Mode
- Automatically supports the main graph/map view desktop-color tinting in Dark Mode
- Draws the separator grid lines in diff views without a "hatched" effect
## Review Notes

Due to parts of the app being unusable when dark mode isn’t supported in that screen, this PR isn’t as small and granular as would be ideal. However, the diff was specifically adjusted to limit unnecessary churn, such as avoiding the automatic rewriting of xibs. I failed to identify good points to break up what remains; if you find chunks you’re rather me split out to review separately, I’d be happy to do so (through the magic of GitUp!).

## Screenshots

<details>
<summary>10.15 (Dark)</summary>
<a target="_blank" rel="noopener noreferrer href="https://user-images.githubusercontent.com/170812/60307735-c892c100-9913-11e9-82f1-e19e67bfdbbc.jpg">10.15 Dark #1</a>
<a target="_blank" rel="noopener noreferrer href="https://user-images.githubusercontent.com/170812/60307736-c892c100-9913-11e9-961d-37aa7740178d.jpg">10.15 Dark #2</a>
<a target="_blank" rel="noopener noreferrer href="https://user-images.githubusercontent.com/170812/60307737-c892c100-9913-11e9-8534-02f0b8db1848.jpg">10.15 Dark #3</a>
<a target="_blank" rel="noopener noreferrer href="https://user-images.githubusercontent.com/170812/60307738-c892c100-9913-11e9-902d-8b2a4cb41ad0.jpg">10.15 Dark #4</a>
<a target="_blank" rel="noopener noreferrer href="https://user-images.githubusercontent.com/170812/60307739-c892c100-9913-11e9-8864-0d35d701e985.jpg">10.15 Dark #5</a>
</details>
<details>
<summary>10.15 (Light)</summary>
<a target="_blank" rel="noopener noreferrer href="https://user-images.githubusercontent.com/170812/60307741-c892c100-9913-11e9-9cea-09783c951437.jpg">10.15 Light #1</a>
<a target="_blank" rel="noopener noreferrer href="https://user-images.githubusercontent.com/170812/60307742-c892c100-9913-11e9-925c-f39e6e79d11e.jpg">10.15 Light #2</a>
<a target="_blank" rel="noopener noreferrer href="https://user-images.githubusercontent.com/170812/60307743-c92b5780-9913-11e9-8fd8-cde92e01e509.jpg">10.15 Light #3</a>
<a target="_blank" rel="noopener noreferrer href="https://user-images.githubusercontent.com/170812/60307744-c92b5780-9913-11e9-9b20-9c2805e1fa01.jpg">10.15 Light #4</a>
<a target="_blank" rel="noopener noreferrer href="https://user-images.githubusercontent.com/170812/60307745-c92b5780-9913-11e9-973b-4117f0c0ab01.jpg">10.15 Light #5</a>
</details>
<details>
<summary>10.14 (Dark)</summary>
<a target="_blank" rel="noopener noreferrer href="https://user-images.githubusercontent.com/170812/60307725-c7619400-9913-11e9-9520-5cf604cbdd69.jpg">10.14 Dark #1</a>
<a target="_blank" rel="noopener noreferrer href="https://user-images.githubusercontent.com/170812/60307726-c7619400-9913-11e9-8b31-0e8b7f57b035.jpg">10.14 Dark #2</a>
<a target="_blank" rel="noopener noreferrer href="https://user-images.githubusercontent.com/170812/60307727-c7619400-9913-11e9-94a6-9a050f4641a7.jpg">10.14 Dark #3</a>
<a target="_blank" rel="noopener noreferrer href="https://user-images.githubusercontent.com/170812/60307728-c7fa2a80-9913-11e9-83f2-49f48dd304f2.jpg">10.14 Dark #4</a>
<a target="_blank" rel="noopener noreferrer href="https://user-images.githubusercontent.com/170812/60307729-c7fa2a80-9913-11e9-82a5-8e9fef62b0bc.jpg">10.14 Dark #5</a>
</details>
<details>
<summary>10.14 (Light)</summary>
<a target="_blank" rel="noopener noreferrer href="https://user-images.githubusercontent.com/170812/60307730-c7fa2a80-9913-11e9-9d88-3167a9b1ea65.jpg">10.14 Light #1</a>
<a target="_blank" rel="noopener noreferrer href="https://user-images.githubusercontent.com/170812/60307731-c7fa2a80-9913-11e9-9bf8-cbb0b33f7681.jpg">10.14 Light #2</a>
<a target="_blank" rel="noopener noreferrer href="https://user-images.githubusercontent.com/170812/60307732-c7fa2a80-9913-11e9-85e3-d51c96e5adf7.jpg">10.14 Light #3</a>
<a target="_blank" rel="noopener noreferrer href="https://user-images.githubusercontent.com/170812/60307733-c7fa2a80-9913-11e9-9d42-5c67af32d71b.jpg">10.14 Light #4</a>
<a target="_blank" rel="noopener noreferrer href="https://user-images.githubusercontent.com/170812/60307734-c7fa2a80-9913-11e9-9de3-f973a8186307.jpg">10.14 Light #5</a>
</details>
<details>
<summary>10.13 (Light)</summary>
<a target="_blank" rel="noopener noreferrer href="https://user-images.githubusercontent.com/170812/60307720-c7619400-9913-11e9-9cd9-cd664b4a9728.jpg"></a>
<a target="_blank" rel="noopener noreferrer href="https://user-images.githubusercontent.com/170812/60307721-c7619400-9913-11e9-9278-d336ad9c9695.jpg"></a>
<a target="_blank" rel="noopener noreferrer href="https://user-images.githubusercontent.com/170812/60307722-c7619400-9913-11e9-8d18-fb8e6447b13f.jpg"></a>
<a target="_blank" rel="noopener noreferrer href="https://user-images.githubusercontent.com/170812/60307723-c7619400-9913-11e9-9541-9b489e0ace0c.jpg"></a>
<a target="_blank" rel="noopener noreferrer href="https://user-images.githubusercontent.com/170812/60307724-c7619400-9913-11e9-9d31-58752c2d6821.jpg"></a>
</details>
<details>
<summary>10.12 (Light)</summary>
<a target="_blank" rel="noopener noreferrer href="https://user-images.githubusercontent.com/170812/60307715-c6c8fd80-9913-11e9-8a0f-fb28c66f7c8f.jpg"></a>
<a target="_blank" rel="noopener noreferrer href="https://user-images.githubusercontent.com/170812/60307716-c6c8fd80-9913-11e9-865b-a6371e627fde.jpg"></a>
<a target="_blank" rel="noopener noreferrer href="https://user-images.githubusercontent.com/170812/60307717-c6c8fd80-9913-11e9-8722-0403f90c801f.jpg"></a>
<a target="_blank" rel="noopener noreferrer href="https://user-images.githubusercontent.com/170812/60307718-c6c8fd80-9913-11e9-9e2f-1cf8046f4b13.jpg"></a>
<a target="_blank" rel="noopener noreferrer href="https://user-images.githubusercontent.com/170812/60307719-c7619400-9913-11e9-8f99-a05bb58b3d9d.jpg"></a>
</details>
<details>
<summary>10.11</summary>
<a target="_blank" rel="noopener noreferrer href="https://user-images.githubusercontent.com/170812/60307710-c6306700-9913-11e9-90c8-86bf7b58f194.jpg"></a>
<a target="_blank" rel="noopener noreferrer href="https://user-images.githubusercontent.com/170812/60307711-c6306700-9913-11e9-98ba-04c3acf0fade.jpg"></a>
<a target="_blank" rel="noopener noreferrer href="https://user-images.githubusercontent.com/170812/60307712-c6306700-9913-11e9-9ffe-30a95ea0e879.jpg"></a>
<a target="_blank" rel="noopener noreferrer href="https://user-images.githubusercontent.com/170812/60307713-c6c8fd80-9913-11e9-990b-eaaec1b49e39.jpg"></a>
<a target="_blank" rel="noopener noreferrer href="https://user-images.githubusercontent.com/170812/60307714-c6c8fd80-9913-11e9-971d-037f4c2424c1.jpg"></a>
</details>
<details>
<summary>10.10</summary>
<a target="_blank" rel="noopener noreferrer href="https://user-images.githubusercontent.com/170812/60307703-c597d080-9913-11e9-9dcd-46491f63e1bf.jpg"></a>
<a target="_blank" rel="noopener noreferrer href="https://user-images.githubusercontent.com/170812/60307704-c597d080-9913-11e9-8d31-62169b093406.jpg"></a>
<a target="_blank" rel="noopener noreferrer href="https://user-images.githubusercontent.com/170812/60307705-c597d080-9913-11e9-9426-ce47cb8251f2.jpg"></a>
<a target="_blank" rel="noopener noreferrer href="https://user-images.githubusercontent.com/170812/60307708-c6306700-9913-11e9-934b-a1d26797f0af.jpg"></a>
<a target="_blank" rel="noopener noreferrer href="https://user-images.githubusercontent.com/170812/60307709-c6306700-9913-11e9-9715-e50c0e9153d6.jpg"></a>
</details>

### And Finally…

I AGREE TO THE GITUP CONTRIBUTOR LICENSE AGREEMENT

### CODE REVIEW

File: GitUp/Application/Document.m [hunk 1]
```diff
@@ -723,14 +727,15 @@ - (void)_updateStatusBar {
     _pullButton.enabled = [_mapViewController validateUserInterfaceItem:(id)_pullButton];
     NSRect frame = _pullButton.frame;
     if (isBehind) {
-      _pullButton.image = [NSImage imageNamed:@"icon_action_fetch_new"];
+      _pullButton.image = _mainWindow.isInDarkAppearance ? [NSImage imageNamed:@"ActionFetchNewDark"] : [NSImage imageNamed:@"ActionFetchNew"];
```
Comment (lucasderraugh) [inline GitUp/Application/Document.m]: Why doesn't NSImage handle this automatically with asset catalog light/dark modes?

File: GitUpKit/Utilities/GIViewController.m [hunk 2]
```diff
@@ -341,6 +360,10 @@ - (void)viewDidFinishLiveResize {
   ;
 }
 
+- (void)viewDidChangeEffectiveAppearance {
+  ;
```
Comment (lucasderraugh) [inline GitUpKit/Utilities/GIViewController.m]: A comment might be nice here to explain why this is being overridden.

**Comment by zwaldowski** [comment]
>  Is this an issue for backwards compatibility that we have to conditionally select the dark/light image in code rather than let the asset catalogs handle this for us?

That is correct.

File: GitUp/Application/Document.m (hunk 1 above)
Comment (zwaldowski) [reply GitUp/Application/Document.m]: See the note in the PR description.

For deploying to 10.8, the compiled asset catalog doesn't contain images at all: it dumps them all with the appropriate names in the Resources directory so AppKit in 10.8 can find them. In this case, the dark representations simply aren't emitted into the final app bundle.

If that were to bump up one, for deploying to 10.9, the images would then all be in the asset catalog, but AppKit in 10.9 doesn't fully understand multiple image representations in the compiled asset catalog and will non-deterministically choose one.

10.10 is the first version to more or less forward-support all the image features of the asset catalog.

File: GitUpKit/Utilities/GIViewController.m (hunk 2 above)
Comment (zwaldowski) [reply GitUpKit/Utilities/GIViewController.m]: Can do. Was just following convention in the file.

**Comment by zwaldowski** [comment]
Will also investigate Travis in a sec…

File: GitUpKit/Utilities/GIViewController.m (hunk 2 above)
Comment (zwaldowski) [reply GitUpKit/Utilities/GIViewController.m]: Addressed in next push

**Comment by zwaldowski** [comment]
Added commentary around `viewDidChangeEffectiveAppearance` et. al. and (I think) fixed Travis.

**Comment by lucasderraugh** [comment]
Seemed like @swisspol was fine with bumping to 10.10 in the original issue. If it would make the commit cleaner to simply support 10.10 and higher I think I would be worth bumping.

**Comment by zwaldowski** [comment]
Will follow-up on that shortly…

**Comment by zwaldowski** [comment]
I will prepare a PR for bumping and then rebase this onto that once merged, or will have it clean up bits of this it can; whichever comes first.

**Comment by zwaldowski** [comment]
OK, it totally is much simpler. I'm getting CI to check on this branch then posting a PR for the deployment change.

**Comment by zwaldowski** [comment]
Rebased now that #534 is merged. Combing over it again myself, hoping to make it just a little bit smaller. Ready for re-review.

File: GitUp/Application/Document.m [hunk 3]
```diff
@@ -750,19 +750,8 @@ - (void)_updateToolBar {
...
-        _snapshotsButton.image = [NSImage imageNamed:@"icon_nav_snapshot_disable"];
-        _snapshotsButton.alternateImage = [NSImage imageNamed:@"icon_nav_snapshot_disable"];
-        _snapshotsButton.enabled = NO;
-      } else if (_snapshotsView.superview) {
-        _snapshotsButton.image = [NSImage imageNamed:@"icon_nav_snapshot_active"];
-        _snapshotsButton.alternateImage = [NSImage imageNamed:@"icon_nav_snapshot_active_pressed"];
-        _snapshotsButton.enabled = YES;
-      } else {
-        _snapshotsButton.image = [NSImage imageNamed:@"icon_nav_snapshot"];
-        _snapshotsButton.alternateImage = [NSImage imageNamed:@"icon_nav_snapshot_pressed"];
```
Comment (lucasderraugh) [inline GitUp/Application/Document.m]: What are we losing with removing these images?

Comment (zwaldowski) [reply GitUp/Application/Document.m]: Nothing. The snapshots button is now configured as a toggle button. With the icon set to a template, it has all the disabled, highlighted, and selected states it did before. As well, the pressed-in appearance follows the system accent color.

**Comment by lucasderraugh** [comment]
Everything looks great. I'm not sure the colors are just right though, it does seem a little too dark if I compare it to Xcode which has slightly softer blacks. I'm not going to block merging on this, but I think we need to add the ability to toggle back to light mode if I'm to push this as an official update. Something like this perhaps:
Screen Shot 2019-07-02 at 10 12 57 PM
If you don't have the time, I'd be happy to add it in when I get the time, but I can imagine some people might not be happy if they can't use it as they were using it before. Pierre's #1 concern was around breaking anything as it works currently, and this could be seen as breaking old behavior. What do you think?

**Comment by zwaldowski** [comment]
>  I'm not sure the colors are just right though, it does seem a little too dark if I compare it to Xcode which has slightly softer blacks. 

This is probably my doing, the map has defaulted to `controlBackgroundColor` like a text view and should probably be something else. Please hold.

> … and this could be seen as breaking old behavior

I am pretty strongly against spending time on adding new code for that until it is requested, but you're welcome to do whatever you want.

Users that this would hypothetically "break" will have already explicitly chosen Dark Mode at the system level. The documentation/HIG/WWDC sessions pretty recommend against custom appearance toggles unless the app is specifically media-focused; otherwise, it is more important to respect the users' choices. Meanwhile, GitUp has been blindingly, searingly white for people who have made that choice for over a year, and so has been certainly "broken" in the same way.

**Comment by swisspol** [comment]
> Pierre's number 1 concern was around breaking anything as it works currently

I was referring to functional behavior. For visuals, if things evolve spatially or in the used colors, I have no problem.

**Comment by lucasderraugh** [comment]
@swisspol Good to know 🙂. @zwaldowski let me know when you're done testing out an alternate dark background color and we'll merge this in.

**Comment by zwaldowski** [comment]
> This is probably my doing, the map has defaulted to `controlBackgroundColor` like a text view and should probably be something else. Please hold.

Edit: This is recommended by the SDK, it's the same color/texture used by all scrolling content. The list of things that use it would be too many to count. In fact, it is the identical counterpart to the hard-coded `whiteColor` used in light mode before this PR. (i.e., `controlBackgroundColor` yields the same pure white in light mode.)

Xcode is not a great point of comparison because it involves custom theming. That said, the "sharpness" might be from the contrast with the branch colors? Ours in this branch are pretty much lifted from their counterparts in Calendar, and those are definitely on the neon side of things. We can definitely evolve those; doing so would be trivial and could happen whenever. In the spirit of moving things forward, if there are still concerns we could open an issue for it and I could experiment with it later this week.

**Comment by lucasderraugh** [comment]
For me it's the red/green/black combination I find a bit jarring, but I'd have to play around with some colors to see what bothers me about it. Let's try it out and see what we think!

**Pull request merged**